### PR TITLE
19.3 fb issue38940

### PR DESCRIPTION
**Pull request by labkey-chrisj**
restores field-level error and warning validation test coverage in Domain Designer.
adds synchronizing wait-for-error/warning methods to domainFieldRow

### CODE REVIEW

File: src/org/labkey/test/tests/DomainDesignerTest.java [hunk 1]
```diff
@@ -305,6 +305,13 @@ public void testBlankNameFieldOnAddedField() throws Exception
         assertTrue("expect error to contain [Please provide a name for each field.] but was[" + hasNoNameError + "]",
                 hasNoNameError.contains("Please provide a name for each field."));
 
+        String warningfieldMessage = noNameRow.setName("&totally not a sketchy sql injection attack")
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/DomainDesignerTest.java]: This text is misleading. We aren't testing anything injection related. Having weird characters in field names just makes scripts harder to write.

File: src/org/labkey/test/tests/DomainDesignerTest.java [hunk 2]
```diff
@@ -305,6 +305,13 @@ public void testBlankNameFieldOnAddedField() throws Exception
...
+                .waitForWarning()
+                .detailsMessage();
+        String expectedWarning = "New field. Warning: SQL queries, R scripts, and other code are easiest to write when field names only contain combination of letters, numbers, and underscores, and start with a letter or underscore";
+        assertTrue("expect error to contain [" + expectedWarning + "] but was[" + warningfieldMessage + "]",
+                warningfieldMessage.contains(expectedWarning));
+
         domainDesignerPage.clickCancelAndDiscardChanges();
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/DomainDesignerTest.java]: Save domain to make sure these are actually non-blocking warnings.

File: src/org/labkey/test/tests/DomainDesignerTest.java [hunk 3]
```diff
@@ -491,15 +498,23 @@ public void testAddFieldsWithReservedNames() throws Exception
         DomainFieldRow clientFieldWarning = domainFormPanel.addField("select * from table");
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/DomainDesignerTest.java]: This test name is inaccurate. We're testing more than just reserved field names.
Something like `testFieldNameErrors` would be more accurate.

File: src/org/labkey/test/tests/DomainDesignerTest.java [hunk 4]
```diff
@@ -491,15 +498,23 @@ public void testAddFieldsWithReservedNames() throws Exception
...
+                modRowDetailsMsg.contains(reservedErrMsg));
+        assertTrue("expect error message to contain [" + blargErrMsg + "] but was [" + blarg1DetailsMsg + "]",
+                blarg1DetailsMsg.contains(blargErrMsg));
+        assertTrue("expect error message to contain [" + blargErrMsg + "] but was [" + blarg2DetailsMsg + "]",
+                blarg2DetailsMsg.contains(blargErrMsg));
+        assertTrue("expect warning message to contain [" + expectedWarnMsg + "] but was [" + clientFieldWarningMsg + "]",
+                clientFieldWarningMsg.contains(expectedWarnMsg));
 
         assertTrue("expect field error when using reserved field names", modifiedRow.hasFieldError());
         assertTrue("expect error for duplicate field names", blarg1.hasFieldError());
```
Comment (labkey-tchad) [inline src/org/labkey/test/tests/DomainDesignerTest.java]: We should verify that we are unable to save the domain if there are errors.

File: src/org/labkey/test/components/domain/DomainFieldRow.java [hunk 5]
```diff
@@ -518,11 +519,27 @@ public boolean hasFieldError()
     {
         return getComponentElement().getAttribute("class").contains("domain-row-border-error");
     }
+    public DomainFieldRow waitForError()
+    {
+        getWrapper().shortWait().ignoring(StaleElementReferenceException.class)
+                .until(driver -> {
+                   return this.hasFieldError();
+                });
+        return this;
```
[the remark this answers is not in the document]
Comment (labkey-chrisj) [reply src/org/labkey/test/components/domain/DomainFieldRow.java]: aww.  Okay.

File: src/org/labkey/test/tests/DomainDesignerTest.java (hunk 2 above)
Comment (labkey-chrisj) [reply src/org/labkey/test/tests/DomainDesignerTest.java]: these are blocking warnings.  It won't enable the 'save' button if your field name is [&totally not a sketchy sql injection attack].  Can't save it that way.

File: src/org/labkey/test/tests/DomainDesignerTest.java (hunk 1 above)
Comment (labkey-chrisj) [reply src/org/labkey/test/tests/DomainDesignerTest.java]: new name: [&has weird characters that make scripts hard to write]

File: src/org/labkey/test/tests/DomainDesignerTest.java (hunk 4 above)
Comment (labkey-chrisj) [reply src/org/labkey/test/tests/DomainDesignerTest.java]: occurs in testBlankNameFieldOnAddedField().

We now also check there to assure that the 'save' button is in fact disabled

File: src/org/labkey/test/tests/DomainDesignerTest.java (hunk 3 above)
Comment (labkey-chrisj) [reply src/org/labkey/test/tests/DomainDesignerTest.java]: I like it.

**Comment by labkey-chrisj** [comment]
Need clarity on why this is being held up

**Pull request merged**